### PR TITLE
[BEAM-27] Add DoFn.OnTimerContext and support as a DoFn parameter

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @amitsela take a look?

**Comment by amitsela** [comment]
@kennknowles happy to take a look, could take me another day or so..

**Comment by kennknowles** [comment]
Cool, thanks for letting me know. I may find someone else for this one, just to unblock because this is my main thread of work right now. I was trying to take advantage of time zones :-)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java [hunk 1]
```diff
@@ -583,6 +583,18 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
   }
 
   /**
+   * Generates a {@link TypeDescriptor} for {@code DoFn<InputT, OutputT>.Context} given {@code
+   * InputT} and {@code OutputT}.
+   */
+  private static <InputT, OutputT>
+      TypeDescriptor<DoFn<InputT, OutputT>.OnTimerContext> doFnOnTimerContextTypeOf(
+          TypeDescriptor<InputT> inputT, TypeDescriptor<OutputT> outputT) {
+    return new TypeDescriptor<DoFn<InputT, OutputT>.OnTimerContext>() {}.where(
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: This formatting seems funky.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java [hunk 2]
```diff
@@ -636,6 +648,20 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
               outputT));
     }
 
+    for (Parameter param : extraParameters) {
+      errors.checkArgument(
+          Predicates.or(
+                  Predicates.instanceOf(Parameter.OnTimerContextParameter.class),
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: Make a constant set and for each check that at least one of `isAssignableFrom()`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java [hunk 3]
```diff
@@ -693,6 +719,24 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
           DoFn.ProcessContext.class.getSimpleName(),
           RestrictionTracker.class.getSimpleName(),
           methodContext.getExtraParameters());
+    } else {
+      // These are the parameters allowed for non-splittable @ProcessElement
+      for (Parameter param : methodContext.getExtraParameters()) {
+        errors.checkArgument(
+            Predicates.or(
+                    Predicates.instanceOf(Parameter.ProcessContextParameter.class),
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: Same validation. Factor into a helper method?

**Comment by kennknowles** [comment]
R: @tgroh -@amitsela (though feel free to comment if you like)

**Comment by kennknowles** [comment]
Yes, indeed. The wiring is up next. Actually already implemented for the most part, rebased on top of this.

**Pull request closed without merging**